### PR TITLE
Adding profiling capability to ddp collective functions

### LINKED ISSUE #43231: Autograd profiler support for torch.distributed
## 🚀 Feature

The [autograd profiler](https://pytorch.org/docs/stable/autograd.html#profiler) does not cover `torch.distributed` ops like allreduce, allgather etc. Adding support for this would be invaluable for debugging performance issues.

We should cover all the collective and point to point operations listed here: https://pytorch.org/docs/stable/distributed.html.

A nice extension for the autograd profiler in the distributed setting would be to have an API like `torch.distributed.combine_profiles`, where a single rank (ex: rank 0) can pull the autograd profiles for all other ranks and provide users with a single chrome trace view which displays the trace across all nodes.

## Motivation

The [autograd profiler](https://pytorch.org/docs/stable/autograd.html#profiler) has been an invaluable tool in PyTorch to debug performance issues and extending this to `torch.distributed` would be beneficial to users. 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @jiayisuse @agolynski

### PR DESCRIPTION
**Pull request by mrzzd**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44436 [WIP] Adding profiling capability to ddp collective functions**

Differential Revision: [D23611359](https://our.internmc.facebook.com/intern/diff/D23611359/)

### CODE REVIEW

**Comment by pritamdamania87** [comment]
Could you update the PR summary to include a sample of what the autograd profiler output looks like with these changes?

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -753,10 +755,35 @@ def recv(tensor,
         return src
 
 
+def async_collective(func):
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:758]: If this isn't meant to be used directly by PT users, maybe prefix it with an `_`?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -753,10 +755,35 @@ def recv(tensor,
...
+    @functools.wraps(func)
+    def wrapped(*args, async_op=False, **kwargs):
+        work = func(*args, **kwargs)
+        if work is None:
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:762]: What are the semantics here? If I call `work.wait()` in a collective and decorate with `@async_collective`, will it not be profiled correctly? If we want to use `async_collective` would the function have to return a `work` object?

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -753,10 +755,35 @@ def recv(tensor,
...
+            return
+
+        should_profile = torch.autograd._profiler_enabled()
+        if should_profile:
+            rf = torch.autograd.profiler.record_function(func.__name__)
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:767]: Should we somehow distinguish between async and sync calls in the profiling output?

**Comment by rohan-varma** [comment]
Thank you for working on this! I was concurrently putting together an RFC for DDP/collectives profiling and hoping to get input from @pritamdamania87 and @ilia-cher, but I will take this PR into account as well while preparing the RFC.

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -753,10 +755,35 @@ def recv(tensor,
...
+        else:
+            rf = contextlib.suppress()
+
+        with rf:
+            if async_op:
+                if should_profile:
+                    fut = rf._call_end_callbacks_on_future(work.get_future())
```
Comment (pritamdamania87) [inline torch/distributed/distributed_c10d.py:774]: `get_future` currently only works for NCCL, I'm wondering if there is away to ensure `with record_function` works for async cuda events.

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (pritamdamania87) [inline torch/distributed/distributed_c10d.py:758]: As @rohan-varma mentioned in a separate thread, we could call `process_group->allreduce()` from C++ (ex: c10d/reducer.cpp). Let's put all the RecordFunction logic in ProcessGroupGloo.cpp and ProcessGroupNCCL.cpp to ensure pure C++ invocations are profiled as well.

File: torch/testing/_internal/distributed/distributed_test.py [hunk 5]
```diff
@@ -914,20 +914,30 @@ def _test_all_reduce_helper(
...
-                        tensor = tensor.cuda(rank_to_GPU[rank][0])
-                    dist.all_reduce(tensor, op, group_id)
-                    self.assertEqual(tensor, _build_tensor(src + 1, expected_value))
+                if cuda:
+                    tensor = tensor.cuda(rank_to_GPU[rank][0])
+                with torch.autograd.profiler.profile() as prof:
+                    work = dist.all_reduce(tensor, op, group_id, async_op=async_op)
+                    if async_op:
+                        work.wait()
+                        work._get_profiling_future().wait()
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:930]: What is the reason to have these 3 lines? Shouldn't the profiling work even without this?

File: torch/lib/c10d/ProcessGroupGloo.hpp [hunk 6]
```diff
@@ -68,20 +68,31 @@ class ProcessGroupGloo : public ProcessGroup {
   //
   class AsyncWork : public ProcessGroup::Work {
    public:
+    AsyncWork() {
+      resultFuture = c10::make_intrusive<c10::ivalue::Future>(
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.hpp:72]: nit: `resultFuture_` for private members.

File: torch/lib/c10d/ProcessGroupGloo.hpp [hunk 7]
```diff
@@ -68,20 +68,31 @@ class ProcessGroupGloo : public ProcessGroup {
...
+    }
+
     static void execute(std::shared_ptr<AsyncWork> work) {
       std::exception_ptr eptr;
       try {
         work->run();
       } catch (...) {
         eptr = std::current_exception();
       }
+      work->resultFuture->markCompleted();
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.hpp:87]: Shouldn't this be done after `work->finish()`?

File: torch/lib/c10d/ProcessGroupGloo.hpp [hunk 8]
```diff
@@ -68,20 +68,31 @@ class ProcessGroupGloo : public ProcessGroup {
   //
   class AsyncWork : public ProcessGroup::Work {
    public:
+    AsyncWork() {
+      resultFuture = c10::make_intrusive<c10::ivalue::Future>(
+          c10::ListType::create(c10::TensorType::get()));
+    }
+
+    c10::intrusive_ptr<c10::ivalue::Future> getFuture() override {
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupGloo.hpp:76]: We should update our docs for `get_future` to reflect this now works for Glooo as well: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L756 (except send/recv).

File: torch/testing/_internal/distributed/distributed_test.py [hunk 9]
```diff
@@ -914,20 +914,30 @@ def _test_all_reduce_helper(
...
+
+                self.assertEqual(tensor, _build_tensor(src + 1, expected_value))
+
+                def get_event(partial_key):
+                    return [event for event in prof.function_events if partial_key in event.name][0]
+
+                recv_event = get_event("all_reduce")
+                self.assertEqual(recv_event.count, 1)
+                self.assertGreater(recv_event.cpu_time, 0)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:939]: We should proobably validate cuda time as well for NCCL?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 10]
```diff
@@ -964,6 +985,25 @@ def test_all_reduce_sum_cuda(self):
...
 
+
+        @require_backends_available({"gloo", "nccl"})
+        @skip_if_no_gpu
+        def test_all_reduce_sum_cuda_async(self):
+            group, group_id, rank = self._init_global_test()
+            rank_to_GPU = self._init_multigpu_helper()
+            self._test_all_reduce_helper(
+                group,
+                group_id,
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:996]: Would be nice to add tests for all the collectives we have enabled to ensure they're profiled correctly.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Awesome work! I think the major change we need is that the profiling should probably be at the C++ level such that C++ invocations are profiled too.

**Comment by mrzzd** [comment]
Closing it in favour of #46471

**Pull request closed without merging**